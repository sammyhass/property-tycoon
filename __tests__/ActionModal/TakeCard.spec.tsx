import ActionModalTakeCard from '@/components/Game/ActionModal/TakeCard';
import { act } from '@testing-library/react';
import React from 'react';
import { fakeCard, fakePlayer, renderWithGameContext } from '../fakers';
jest.useFakeTimers();

describe('Take Card Action Modal', () => {
  describe('Opportunity Knocks', () => {
    it('matches snapshot', () => {
      const { container } = renderWithGameContext(
        <ActionModalTakeCard cardType="OPPORTUNITY_KNOCKS" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('shows correct text', () => {
      const { container } = renderWithGameContext(
        <ActionModalTakeCard cardType="OPPORTUNITY_KNOCKS" />
      );
      expect(container.textContent).toContain('Opportunity Knocks');
    });

    it('card can be taken once button clicked and action is performed after next click', () => {
      const card = fakeCard({
        type: 'OPPORTUNITY_KNOCKS',
        action_type: 'PAY_BANK',
        description: 'Pay bank £200',
      });
      const takeCard = jest.fn().mockReturnValue(card);
      const performCardAction = jest.fn();
      const { container } = renderWithGameContext(
        <ActionModalTakeCard cardType="OPPORTUNITY_KNOCKS" />,
        {
          takeCard,
          performCardAction,
          currentPlayer: fakePlayer('cat'),
        }
      );

      const takeCardButton = container.querySelector('button');

      act(() => {
        takeCardButton?.click();
        jest.runAllTimers();
      });
      expect(takeCard).toHaveReturnedWith(card);
      expect(container.textContent).toContain(card.description);

      // Take Card Button now becomes the 'action' button
      act(() => {
        takeCardButton?.click();
        jest.runAllTimers();
      });

      expect(performCardAction).toHaveBeenCalledWith('cat', card);
    });
  });

  describe('Pot Luck', () => {
    it('matches snapshot', () => {
      const { container } = renderWithGameContext(
        <ActionModalTakeCard cardType="POT_LUCK" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('shows correct text', () => {
      const { container } = renderWithGameContext(
        <ActionModalTakeCard cardType="POT_LUCK" />
      );
      expect(container.textContent).toContain('Pot Luck');
    });

    it('card can be taken once button clicked and action is performed after next click', () => {
      const card = fakeCard({
        type: 'POT_LUCK',
        action_type: 'PAY_BANK',
        description: 'Pay bank £200',
      });
      const performCardAction = jest.fn();
      const takeCard = jest.fn().mockReturnValue(card);
      const { container } = renderWithGameContext(
        <ActionModalTakeCard cardType="POT_LUCK" />,
        {
          takeCard: takeCard,
          performCardAction,
          currentPlayer: fakePlayer('cat'),
        }
      );

      const takeCardButton = container.querySelector('button');

      act(() => {
        takeCardButton?.click();
        jest.runAllTimers();
      });

      expect(container.textContent).toContain(card.description);
      expect(takeCard).toHaveReturnedWith(card);

      // Take Card Button now becomes the 'action' button
      act(() => {
        takeCardButton?.click();
        jest.runAllTimers();
      });

      expect(performCardAction).toHaveBeenCalledWith('cat', card);
    });
  });
});
