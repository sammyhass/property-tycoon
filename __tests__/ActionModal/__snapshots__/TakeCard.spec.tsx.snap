// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Take Card Action Modal Opportunity Knocks matches snapshot 1`] = `
<div
  class="css-u6v8er"
>
  <h2
    class="chakra-heading css-0"
  >
    Pick a 
    Opportunity Knocks
     
    Card
  </h2>
  <button
    class="chakra-button css-lb8ny5"
    type="button"
  >
    Take
  </button>
</div>
`;

exports[`Take Card Action Modal Pot Luck matches snapshot 1`] = `
<div
  class="css-u6v8er"
>
  <h2
    class="chakra-heading css-0"
  >
    Pick a 
    Pot Luck
     
    Card
  </h2>
  <button
    class="chakra-button css-lb8ny5"
    type="button"
  >
    Take
  </button>
</div>
`;
